tests/cpp/test_multidevice_tutorial.cpp

```diff
@@ -0,0 +1,367 @@
+    std::cout << tv0->toString() << std::endl;
+  }
+  // Alternatively:
+  tv0->setDeviceMesh(mesh_full);
+  // means that tv0 is replicated on all the devices.
+  if (verbose_print) {
+    std::cout << tv0->toString() << std::endl;
+  }
+
+  // Let us exercise this notion by examining two simple situations where we set
+  // the device mesh. In those simple examples, no network communication will be
+  // needed at all. We will simply see that only through setting the device mesh
+  // we can, e.g., replicate computation on all devices or select a subset of
+  // devices.
+
+  // FULL REPLICATION ON ALL DEVICES
+  // Let us set tv0 and tv1 to be replicated on all the devices:
+  tv0->setDeviceMesh(mesh_full);
+  tv1->setDeviceMesh(mesh_full);
+
+  // RUNTIME
+  // Set up the input
+  constexpr int64_t tensor_size = 128;
+  const c10::TensorOptions tensor_options =
+      at::TensorOptions().device(communicator->device()).dtype(at::kFloat);
+  // each rank allocate a tensor a on different device
+  at::Tensor input = at::randn({tensor_size}, tensor_options);
+  {
+    // EXECUTION
+    MultiDeviceExecutor multidevice_executor(
+        std::make_unique<Fusion>(fusion), *communicator);
+    if (verbose_print) {
+      fusion.printMath();
+      fusion.printKernel();
+      multidevice_executor.print();
+    }
+
+    // The same computation is replicated on all the devices
+    at::Tensor output = multidevice_executor.runWithInput({input}).at(0);
+
+    // VALIDATION
+    // Each device produces the full output
+    EXPECT_TRUE(output.equal(input));
+  }
+
+  // SINGLE DEVICE EXECUTION
+  // This indicates that tv0 and tv1 are present on device 0 only
+  tv0->setDeviceMesh(mesh_zero);
+  tv1->setDeviceMesh(mesh_zero);
+  // RUNTIME
+  {
+    // EXECUTION
+    MultiDeviceExecutor multidevice_executor(
+        std::make_unique<Fusion>(fusion), *communicator);
+    // here, the compute is done on device 0 only. Other devices don't even read
+    // the input.
+    at::Tensor output = multidevice_executor.runWithInput({input}).at(0);
+
+    // VALIDATION
+    // Only device 0 receives a non-void output
+    if (communicator->deviceId() == 0) {
+      EXPECT_TRUE(output.equal(input));
+    } else {
+      EXPECT_EQ(output.numel(), 0);
+    }
+  }
+}
+
+// Let us now complexify the above example by showing how setting the tensor's
+// device mesh can encode pipelining using multiple devices. Let us define a
+// Fusion which will be executed in a two-stage pipeline fashion. Executing this
+// pipeling will require a network communication between the two pipeline
+// Stages.
+TEST_F(MultiDeviceTutorial, SimplePipelining) {
+  constexpr int64_t tensor_size = 128;
+
+  Communicator* const communicator = communicator_;
+  if (communicator->is_available() == false || communicator->size() < 2) {
+    GTEST_SKIP() << "distributed setting is not available";
+  }
+  if (communicator->deviceId() > 1) { // we only need 2 devices for this test
+    return;
```

Comment:
Is this necessary? What would happen if the other processes kept running the rest of the test?

Reply:
it is not necessary. I used it to make a less verbose print but I can move this predicate to where printing happens.